### PR TITLE
Move helper into painless

### PR DESCRIPTION
**Pull request by nik9000**
This moves `TemporalAccessor#toEpochMillis` out of runtime fields and
into painless directly so all contexts can use it. Now something like:
```
ZonedDateTime.parse('2020-10-07T19:02:56.782Z').toEpochMilli()
```
is possible.

### CODE REVIEW

**Comment by nik9000** [comment]
@jdconrad I checked on whether or not I could have augmented the core type in runtime fields and it looks like I can. When I tried a few months ago I must have done something wrong.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 1]
```diff
@@ -712,4 +715,13 @@ public static Matcher matcher(Pattern receiver, int limitFactor, CharSequence in
...
         return receiver.matcher(new LimitedCharSequence(input, receiver, limitFactor));
     }
+
+    /**
+     * Convert a {@link TemporalAccessor} into millis since epoch like {@link Instant#toEpochMilli()}.
+     */
+    public static long toEpochMilli(TemporalAccessor v) {
+        long millis = Math.multiplyExact(v.getLong(ChronoField.INSTANT_SECONDS), 1_000);
+        millis = Math.addExact(millis, v.get(ChronoField.NANO_OF_SECOND) / 1_000_000);
+        return millis;
```
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:725]: No suggestions here. Just surprised this is the best way to do it :).

**Review by jdconrad — APPROVED** [review]
LGTM. Thanks for adding this.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 1 above)
Comment (nik9000) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:725]: Its not exactly the clearest thing, no. If you *know* that you have a `ZonedDateTime` you can do `zdt.toInstant().toEpochMilli()` which does about the same thing but you have to worry about if the JVM can skip allocating the `Instant`.... And we're not sure we have a `ZonedDateTime` here either.

Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:725]: Makes sense. Thanks for the explanation!

Comment (rjernst) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:725]: Why the `Math.` calls instead of actual math?

**Comment by nik9000** [comment]
Just so we don't silently overflow. It'd be a nasty surprise if your huge
date flips to small accidentally.

**Comment by nik9000** [comment]
Looks like it'd roll over 292 million years from now. So this is only gonna
come up with folks trying to do astronomy stuff, I guess.

**Comment by rjernst** [comment]
> this is only gonna come up with folks trying to do astronomy stuff

This doesn't seem like a good reason to use slower arithmetic? Also, we have other places in ES that use milli/nano math and don't check for overflow.

**Comment by nik9000** [comment]
> This doesn't seem like a good reason to use slower arithmetic? Also, we have other places in ES that use milli/nano math and don't check for overflow.

Ultimately I did it because it is similar to what java.util.time does, but they do tend to be more defensive than we are. I'll switch it.

**Comment by nik9000** [comment]
@rjernst I swapped it out to regular math. Let me know what you think!

**Review by rjernst — APPROVED** [review]
LGTM. Thanks Nik.

**Comment by nik9000** [comment]
Thanks folks!

**Pull request merged**